jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java

```diff
@@ -0,0 +1,735 @@
+        }
+        return maxElement;
+    }
+
+    // node build
+
+    private V createVector2D(String line)
+    {
+        String[] elements = splitLineElements(line, 3);
+        // index of the point in elements[0] is not necessary
+        double x = Double.parseDouble(elements[1]);
+        double y = Double.parseDouble(elements[2]);
+        return vectorFactory.apply(new double[] { x, y });
+    }
+
+    private V createVector3D(String line)
+    {
+        String[] elements = splitLineElements(line, 4);
+        // index of the point in elements[0] is not necessary
+        double x = Double.parseDouble(elements[1]);
+        double y = Double.parseDouble(elements[2]);
+        double z = Double.parseDouble(elements[3]);
+        return vectorFactory.apply(new double[] { x, y, z });
+    }
+
+    private static String[] splitLineElements(String line, int expectedLineElements)
+    {
+        String[] elements = line.split(" ");
+        if (elements.length != expectedLineElements) {
+            throw new IllegalArgumentException(
+                "Unexpected number of line elements: " + elements.length + " in line: " + line);
+        }
+        return elements;
+    }
+
+    // read of coordinates and graph computation
+
+    /**
+     * Read the complete {@link Graph} of node locations and returns true if all locations are
+     * distinct.
+     */
+    private boolean readCompleteNodeGraphData(
+        Graph<V, DefaultWeightedEdge> graph, BufferedReader reader,
+        Function<String, V> vectorFactory, ToIntBiFunction<V, V> edgeWeightFunction,
+        Integer dimension)
+        throws IOException
+    {
+        List<V> nodeCoordinates = readNodeCoordinateData(reader, vectorFactory, dimension);
+
+        buildCompleteGraph(graph, nodeCoordinates, edgeWeightFunction);
+        return graph.vertexSet().size() == nodeCoordinates.size();
+    }
+
+    private List<V> readNodeCoordinateData(
+        BufferedReader reader, Function<String, V> vectorFactory, Integer dimension)
+        throws IOException
+    {
+        List<V> coordinates = dimension != null ? new ArrayList<>(dimension) : new ArrayList<>();
+
+        for (String line; (line = reader.readLine()) != null;) {
+            String lineContent = line.trim();
+            if ("EOF".equals(lineContent)) {
+                break;
+            }
+            coordinates.add(vectorFactory.apply(lineContent));
+        }
+        return coordinates;
+    }
+
+    private static <V, E> void buildCompleteGraph(
+        Graph<V, E> graph, List<V> locations, ToIntBiFunction<V, V> edgeWeightFunction)
+    {
+        locations.forEach(graph::addVertex);
+
+        new CompleteGraphGenerator<V, E>().generateGraph(graph, null);
+
+        // Compute edge weights:
+        // Enable parallel streams. Synchronization not necessary, because DefaultWeightedEdge are
+        // IntrusiveEdges and modified directly.
+
+        graph.edgeSet().parallelStream().forEach(e -> {
```

Comment:
Remove the parallel stream here, in order to also avoid any problems with using a generic edge type `E`. 

In the last few years, we explicitly try to avoid code which makes assumptions about the user edge.

Reply:
Moreover, these are probably small instances since the TSP is a hard problem, giving no real motivation for parallel processing.